src/utils/dlck/dlck_engine.c

```diff
@@ -0,0 +1,640 @@
+	}
+
+	*engine_ptr = engine;
+
+	return 0;
+
+fail_vos_fini:
+	vos_db_fini();
+fail_vos_tls_fini:
+	vos_standalone_tls_fini();
+fail_tls_key_delete:
+	ds_tls_key_delete();
+fail_vos_sm_fini:
+	(void)vos_srv_module.sm_fini();
+fail_unregister_keys:
+	dss_unregister_key(&vos_module_key);
+	dss_unregister_key(&daos_srv_modkey);
+	bio_nvme_fini();
+fail_abt_fini:
+	(void)dlck_abt_fini(engine);
+fail_engine_free:
+	dlck_engine_free(engine);
+
+	return rc;
+}
+
+int
+dlck_engine_stop(struct dlck_engine *engine)
+{
+	int rc;
+
+	rc = xstream_stop_all(engine);
+	if (rc != DER_SUCCESS) {
+		/** not all execution streams were stopped - can't pull out other resources */
+		return rc;
+	}
+
+	vos_db_fini();
+
+	vos_standalone_tls_fini();
+
+	ds_tls_key_delete();
+
+	rc = vos_srv_module.sm_fini();
+	if (rc != DER_SUCCESS) {
+		/** this is odd - do not free other resources just in case */
+		return rc;
+	}
+
+	dss_unregister_key(&vos_module_key);
+	dss_unregister_key(&daos_srv_modkey);
+
+	bio_nvme_fini();
+
+	rc = dlck_abt_fini(engine);
+
+	dlck_engine_free(engine);
+
+	return rc;
+}
+
+int
+dlck_engine_exec_all(struct dlck_engine *engine, dlck_ult_func exec_one,
+		     arg_alloc_fn_t arg_alloc_fn, void *custom, arg_free_fn_t arg_free_fn)
+{
+	struct dlck_ult *ults;
+	void           **ult_args;
+	int              rc;
+	int              rc2;
+
+	D_ALLOC_ARRAY(ults, engine->targets);
+	if (ults == NULL) {
+		return -DER_NOMEM;
+	}
+
+	D_ALLOC_ARRAY(ult_args, engine->targets);
+	if (ult_args == NULL) {
+		D_FREE(ults);
+		return -DER_NOMEM;
+	}
+
+	for (int i = 0; i < engine->targets; ++i) {
+		/** prepare arguments */
+		rc = arg_alloc_fn(engine, i, custom, &ult_args[i]);
+		if (rc != DER_SUCCESS) {
+			goto fail_join_and_free;
+		}
+
+		/** start an ULT */
+		rc = dlck_ult_create(engine->xss[i].pool, exec_one, ult_args[i], &ults[i]);
+		if (rc != DER_SUCCESS) {
+			goto fail_join_and_free;
+		}
+	}
+
+	for (int i = 0; i < engine->targets; ++i) {
+		rc = ABT_thread_join(ults[i].thread);
+		if (rc != ABT_SUCCESS) {
+			rc = dss_abterr2der(rc);
+			goto fail_join_and_free;
+		}
+
+		rc = ABT_thread_free(&ults[i].thread);
+		if (rc != ABT_SUCCESS) {
+			rc = dss_abterr2der(rc);
+			goto fail_join_and_free;
+		}
+
+		rc = arg_free_fn(custom, &ult_args[i]);
+		if (rc != 0) {
+			goto fail_join_and_free;
+		}
+	}
+
+	D_FREE(ult_args);
+	D_FREE(ults);
+
+	return DER_SUCCESS;
+
+fail_join_and_free:
+	for (int i = 0; i < engine->targets; ++i) {
+		if (ults[i].thread != ABT_THREAD_NULL) {
+			rc2 = ABT_thread_join(ults[i].thread);
+			if (rc2 != ABT_SUCCESS) {
+				/**
+				 * the ULT did not join - can't free the thread nor free the
+				 * arguments
+				 */
+				continue;
+			}
+		}
+		(void)ABT_thread_free(&ults[i].thread);
+		(void)arg_free_fn(custom, &ult_args[i]);
+	}
+
+	D_FREE(ult_args);
+	D_FREE(ults);
+
+	return rc;
+}
+
+int
+dlck_abt_pool_open(ABT_mutex mtx, const char *storage_path, uuid_t po_uuid, int tgt_id,
+		   daos_handle_t *poh)
```

Comment:
The function name indicating an "ABT pool open", but it actually open a VOS pool?

Reply:
Sorry. I just realized how confusing this is. It is a thread-safe version of `dlck_pool_open()` which opens a VOS pool. Let's call it `dlck_pool_open_safe()`. 😁